enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go

```diff
@@ -6,92 +6,218 @@ import (
+	remote := env.GetByteStreamClient()
+	if remote == nil {
+		return nil, fmt.Errorf("A remote ByteStreamClient is required to enable ByteStreamServerProxy")
 	}
 	return &ByteStreamServerProxy{
-		env:         env,
-		localCache:  localCache,
-		remoteCache: remoteCache,
+		env:    env,
+		local:  local,
+		remote: remote,
 	}, nil
 }
 
 func (s *ByteStreamServerProxy) Read(req *bspb.ReadRequest, stream bspb.ByteStream_ReadServer) error {
-	remoteStream, err := s.remoteCache.Read(stream.Context(), req)
+	localReadStream, err := s.local.Read(stream.Context(), req)
 	if err != nil {
-		return err
+		return s.readRemote(req, stream)
 	}
+	responseSent := false
 	for {
-		rsp, err := remoteStream.Recv()
+		rsp, err := localReadStream.Recv()
 		if err == io.EOF {
 			break
 		}
 		if err != nil {
+			// If some responses were streamed to the client, just return the
+			// error. Otherwise, fall-back to remote.
+			if responseSent {
+				return err
+			} else {
+				return s.readRemote(req, stream)
+			}
+		}
+
+		if err := stream.Send(rsp); err != nil {
+			return err
+		}
+		responseSent = true
+	}
+	return nil
+}
+
+// The Write() RPC requires the client keep track of some state. This struct
+// and its methods take care of that.
+type localWriter struct {
+	ctx          context.Context
+	local        bspb.ByteStream_WriteClient
+	resourceName string
+	initialized  bool
+	offset       int64
+}
+
+func (s *localWriter) send(data []byte) error {
+	req := &bspb.WriteRequest{WriteOffset: s.offset, Data: data}
+	if !s.initialized {
+		// Read resources have a different name format than written resources
+		// Re-write the resource name with a fake execution ID so it passes
+		// the write regex.
+		req.ResourceName = "/uploads/00000000-0000-0000-0000-000000000000" + s.resourceName
+		s.initialized = true
+	}
+	s.offset += int64(len(data))
+	return s.local.Send(req)
+}
+
+func (s *localWriter) close() error {
+	if err := s.local.Send(&bspb.WriteRequest{WriteOffset: s.offset, FinishWrite: true}); err != nil {
+		return err
+	}
+	// Ignore the local response (but not the error)
+	_, err := s.local.CloseAndRecv()
+	return err
+}
+
+func (s *ByteStreamServerProxy) readRemote(req *bspb.ReadRequest, stream bspb.ByteStream_ReadServer) error {
+	remoteReadStream, err := s.remote.Read(stream.Context(), req)
+	if err != nil {
+		return err
+	}
+
+	var localWriteStream *localWriter = nil
+	if req.ReadOffset == 0 {
+		localStream, err := s.local.Write(stream.Context())
+		if err != nil {
+			log.Warningf("error writing to local bytestream server: %s", err)
+		}
+		localWriteStream = &localWriter{
+			ctx:          stream.Context(),
+			local:        localStream,
+			resourceName: req.ResourceName,
+			initialized:  false,
+			offset:       int64(0),
+		}
+	}
+
+	for {
+		rsp, err := remoteReadStream.Recv()
+		if err != nil {
+			if localWriteStream != nil {
+				localWriteStream.close()
```

Comment:
should this function be called commit instead of close()? even though this is an error case, close() will attempt to finish the write. the write would fail, but it doesn't seem like we should be trying to commit it in the first place on error.

Reply:
renamed and restructured to not call on errors except EOF